src/code_translation/ir/nodes.py

```diff
@@ -176,82 +217,61 @@ def sql_string(self) -> Optional[str]:
+        right_on: list[str]
+        if self.right_on == "key":
+            assert isinstance(self.right, SetKeyNode)  # TODO: Don't do that
+            right_on = self.right.key
         else:
-            return None
+            assert isinstance(self.right_on, list)  # this is fine
+            right_on = self.right_on
 
         # Build query string
-        if left_set_key and right_set_key:
-            # TODO: DataFrameNode needs key also aka 'index_col' in read_sql
-            left_table_alias = "S1"
-            right_table_alias = "S2"
-            return f"SELECT * FROM ({left_sql}) AS {left_table_alias} {join_operator} ({right_sql}) AS {right_table_alias} ON {left_table_alias}.{left_key} = {right_table_alias}.{right_key}"
-        elif self.on:
-            left_table_alias = "S1"
-            right_table_alias = "S2"
-            return f"SELECT * FROM ({left_sql}) AS {left_table_alias} {join_operator} ({right_sql}) AS {right_table_alias} ON {left_table_alias}.{self.on} = {right_table_alias}.{self.on}"
-        else:
-            return f"SELECT * FROM ({left_sql}) {join_operator} ({right_sql})"
-
-    def to_cst_translation(self, sql_access_method) -> CSTTranslation:
-        query_str = '"' + self.sql_string + '"'
+        # TODO: DataFrameNode needs key also aka 'index_col' in read_sql
+        left_table_alias = "S1"
+        right_table_alias = "S2"
+        return f"SELECT * FROM ({left_sql}) AS {left_table_alias} {join_operator} ({right_sql}) AS {right_table_alias} ON {' AND '.join(f'{left_table_alias}.{left} = {right_table_alias}.{right}' for (left, right) in zip(left_on, right_on))}"
 
-        # Build cst.Assign Object
-        func = sql_access_method
-        # Assume both left and right have the same con
-        args = [cst.Arg(value=cst.SimpleString(value=query_str)), cst.Arg(value=self.con)]
-
-        return CSTTranslation(code=cst.Call(func=func, args=args))
+    @property
+    def tempVars(self) -> list[TempVar]:
+        return list(set(self.left.tempVars + self.right.tempVars))
 
 
 class SetKeyNode(DataFrameNode):
-    def __init__(self, node: DataFrameNode, key: cst.CSTNode, *args, **kwargs):
+    def __init__(self, node: DataFrameNode, key: list[str], *args, **kwargs):
         node.parent = self
 
         self.node = node
         self.key = key
 
         super().__init__(*args, **kwargs)
 
-    def to_cst_translation(self) -> CSTTranslation:
-        func = cst.Attribute(value=self.node.to_cst_statement(), attr="join")
-        args = [cst.Arg(value=self.key)]
+    @property
+    def sql_string(self) -> str:
+        return self.node.sql_string
+
+    @property
+    def tempVars(self) -> list[TempVar]:
+        return self.node.tempVars
+
+    @property
+    def columns(self) -> list[str] | TempVar:
+        return self.node.columns
 
-        return CSTTranslation(code=cst.Call(func=func, args=args))
+    @property
+    def con(self) -> cst.Name:
+        return self.node.con
```

Comment:
Es macht vielleicht sinn eine property defaultChild oder so hinzuzufügen und dann diese methoden in der basisklasse implementieren zu können. weniger code